### PR TITLE
Chalice invoke command

### PR DESCRIPTION
**Pull request by stealthycoin**
Implements #900 

### Sample app
```python
from chalice import Chalice

app = Chalice(app_name='whatever')


@app.route('/')
def index():
    return {'hello': 'world'}


@app.lambda_function()
def error(event, context):
    return context


@app.lambda_function()
def echo(event, context):
    return event
```

### Calling a lambda function
Providing a payload is done from stdin
```
$ echo '"hello world"' | chalice invoke -n echo
"hello world"
```

No payload
```
$ chalice invoke -n echo
{}
```

### Calling a route
You can call a route by using the `api_handler` function name and fake all the values that API Gateway would send. This is a little nosier:
```
$ cat apig.json
{
    "queryStringParameters": [],
    "headers": {},
    "pathParameters": {},
    "body": {},
    "stageVariables": {},
    "requestContext": {
        "resourcePath": "/",
        "httpMethod": "GET"
    }
}
$ cat apig.json | chalice invoke -n api_handler
{"headers": {}, "statusCode": 200, "body": "{\"hello\": \"world\"}"}
```

### Unhandled errors will show a stack trace
Errors will show a formatted stack trace to look like a local python function failed. This prevents the need to wait for cloudwatch logs to propagate.
```
$ chalice invoke -n error
Traceback (most recent call last):
  File "/var/runtime/awslambda/bootstrap.py", line 110, in decimal_serializer
    raise TypeError(repr(o) + " is not JSON serializable")
  File "/var/lang/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/var/lang/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/var/lang/lib/python3.6/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
TypeError: <__main__.LambdaContext object at 0x7f3fa4d23f28> is not JSON serializable
```

### Handled errors are printed out
Errors thrown by the invoke command itself will be printed out in a pretty format. For example providing a non-json payload causes the `InvalidRequestContentException` to be thrown by lambda.
```
$ echo "invalid json" | chalice invoke -n echo
Error occured when invoking lambda: InvalidRequestContentException
Could not parse request body into json: Unrecognized token 'invalid': was expecting ('true', 'false' or 'null')
 at [Source: [B@30c3127a; line: 1, column: 9]
Aborted!
```

### CODE REVIEW

File: chalice/cli/__init__.py [hunk 1]
```diff
@@ -182,6 +189,59 @@ def deploy(ctx, autogen_policy, profile, api_gateway_stage, stage,
...
+                    'This is the logical name of the function. The only '
+                    'exception is to invoke one of the routes the name '
+                    'api_handler should be used.'))
+@click.option('--profile', metavar='PROFILE',
+              help='Override profile at deploy time.')
+@click.option('--stage', metavar='STAGE', default=DEFAULT_STAGE_NAME,
+              help=('Name of the Chalice stage to deploy to. '
+                    'Specifying a new chalice stage will create '
+                    'an entirely new set of AWS resources.'))
+@click.option('--connection-timeout',
```
Comment (kyleknap) [inline chalice/cli/__init__.py]: We should just read this from the `.chalice/config.json` instead

File: chalice/cli/__init__.py [hunk 2]
```diff
@@ -182,6 +189,59 @@ def deploy(ctx, autogen_policy, profile, api_gateway_stage, stage,
...
+@cli.command('invoke')
+@click.option('-n', '--name', metavar='NAME', required=True,
+              help=('The name of the function to invoke. '
+                    'This is the logical name of the function. The only '
+                    'exception is to invoke one of the routes the name '
+                    'api_handler should be used.'))
+@click.option('--profile', metavar='PROFILE',
+              help='Override profile at deploy time.')
+@click.option('--stage', metavar='STAGE', default=DEFAULT_STAGE_NAME,
+              help=('Name of the Chalice stage to deploy to. '
```
Comment (kyleknap) [inline chalice/cli/__init__.py]: Probably should reword the `--stage` help command here. Looks like this is from the `deploy` command.

File: chalice/awsclient.py [hunk 3]
```diff
@@ -179,6 +186,25 @@ def _is_settling_error(self, error):
...
+            'InvocationType': 'RequestResponse',
+        }
+        if context is not None:
+            kwargs['ClientContext'] = context
+        if payload is not None:
+            kwargs['Payload'] = payload
+
+        try:
+            return self._call_client_method_with_retries(
+                self._client('lambda').invoke, kwargs, max_attempts=1
```
Comment (kyleknap) [inline chalice/awsclient.py]: Is there a specific reason we need `max_retries` on this?

File: chalice/awsclient.py [hunk 4]
```diff
@@ -812,7 +838,7 @@ def _add_lambda_permission_if_needed(self, source_arn, function_arn,
         policy = self.get_function_policy(function_arn)
         if self._policy_gives_access(policy, source_arn, service_name):
             return
-        random_id = self._random_id()
+        random_id = generate_random_id()
```
Comment (kyleknap) [inline chalice/awsclient.py]: Let's keep this an internal method unless some other module uses the utility function and find some other way to get the file under 1000 lines.

File: chalice/cli/__init__.py [hunk 5]
```diff
@@ -182,6 +189,59 @@ def deploy(ctx, autogen_policy, profile, api_gateway_stage, stage,
     reporter.display_report(deployed_values)
 
 
+@cli.command('invoke')
+@click.option('-n', '--name', metavar='NAME', required=True,
+              help=('The name of the function to invoke. '
+                    'This is the logical name of the function. The only '
```
Comment (kyleknap) [inline chalice/cli/__init__.py]: Maybe rephrase the second sentence to: If the function is decorated by `app.route`, use `api_handler` as the name instead

File: chalice/invoke.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    _TRACEBACK_HEADING = 'Traceback (most recent call last):\n'
+
+    def __init__(self, ui):
+        # type: (UI) -> None
+        self._ui = ui
+
+    def format_response(self, response):
+        # type: (Dict[str, Any]) -> None
+        payload = response[self._PAYLOAD_KEY].read()
+        if self._ERROR_KEY in response and \
```
Comment (kyleknap) [inline chalice/invoke.py]: Can be refactored to:

`if response.get(self._ERROR_KEY) == self._ERROR_VALUE`

File: chalice/awsclient.py (hunk 3 above)
Comment (stealthycoin) [reply chalice/awsclient.py]: Yes its a new required parameter of that method after I rebased it.

File: chalice/awsclient.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
 this class to get improved type checking across chalice.
 
 """
+# pylint: disable=too-many-lines
```
Comment (kyleknap) [inline chalice/awsclient.py]: We can probably remove this line. This is under 1000 lines

File: tests/functional/cli/test_cli.py [hunk 8]
```diff
@@ -354,3 +356,52 @@ def test_can_provide_lambda_name_for_logs(runner, mock_cli_factory):
...
+    invoke_handler = mock.Mock(spec=LambdaInvokeHandler)
+    mock_cli_factory.create_lambda_invoke_handler.return_value = invoke_handler
+    mock_cli_factory.create_payload_from_stdin.return_value = 'barbaz'
+    with runner.isolated_filesystem():
+        cli.create_new_project_skeleton('testproject')
+        os.chdir('testproject')
+        result = _run_cli_command(runner, cli.invoke, ['-n', 'foo'],
+                                  cli_factory=mock_cli_factory)
+        assert result.exit_code == 0
+    assert invoke_handler.invoke.call_args == mock.call('barbaz')
```
Comment (kyleknap) [inline tests/functional/cli/test_cli.py]: I wonder if it makes sense to assert the output here in the functional tests. Not sure how feasible it is, but it seems to worthwhile to look at more than just the return code for these tests.

**Comment by kyleknap** [comment]
We should probably add it as a ``feature`` in the changelog as well.

File: chalice/awsclient.py [hunk 9]
```diff
@@ -179,6 +186,19 @@ def _is_settling_error(self, error):
             return True
         return False
 
+    def invoke_function(self, name, context=None, payload=None):
```
Comment (jamesls) [inline chalice/awsclient.py]: What's the `context` param used for?

File: chalice/awsclient.py [hunk 10]
```diff
@@ -179,6 +186,19 @@ def _is_settling_error(self, error):
...
+        # type: (str, _OPT_STR, bytes) -> Dict[str, Any]
+        kwargs = {
+            'FunctionName': name,
+            'InvocationType': 'RequestResponse',
+        }
+        if payload is not None:
+            kwargs['Payload'] = payload
+
+        return self._call_client_method_with_retries(
+            self._client('lambda').invoke, kwargs, max_attempts=1
```
Comment (jamesls) [inline chalice/awsclient.py]: I might have missed this in the review comments but what's the point of using `_call_client_method_with_retries` with a retry of 1?

File: chalice/cli/factory.py [hunk 11]
```diff
@@ -157,6 +187,39 @@ def create_log_retriever(self, session, lambda_arn):
...
+        try:
+            resource = deployed.resource_values(name)
+            arn = resource['lambda_arn']
+        except (KeyError, ValueError):
+            raise NoSuchFunctionError(name)
+
+        function_scoped_config = config.scope(stage, name)
+        session = self.create_botocore_session(
+            read_timeout=function_scoped_config.lambda_timeout,
+            max_retries=0,
```
Comment (jamesls) [inline chalice/cli/factory.py]: Can you add a comment about why we set `max_retries=0`.  It may not be obvious to others why we want to disable retries when invoking the lambda function.

File: chalice/awsclient.py (hunk 9 above)
Comment (stealthycoin) [reply chalice/awsclient.py]: Missed removing it. The first POC had the ability to pass a custom context as well.

File: chalice/awsclient.py (hunk 10 above)
Comment (stealthycoin) [reply chalice/awsclient.py]: Is there an alternative? The method I was using didn't have that but got removed when I rebased. It looked like the expectation was to use that method. I don't really see a reason to re-write the old method when it was deleted in favor of this one.

Comment (jamesls) [reply chalice/awsclient.py]: >Is there an alternative?

I still like I'm missing something, but I'm specifically asking what's the reason for not just using `self._client('lambda').invoke(**kwargs)`?

The call client method with retries is for retries, and if you have a `max_attempts` of 1 that's effectively a noop is it not?

> It looked like the expectation was to use that method. 

I touched on this in my previous comment:

> That method was added specifically because there's some settling time when making config changes (roles, event sources). We shouldn't need this here.


If there's not a good reason to use it, I'd prefer to just keep the code simpler.
